wpilibjJavacPlugin/src/main/java/org/wpilib/javacplugin/ReturnValueUsedListener.java

```diff
@@ -0,0 +1,173 @@
+
+    @Override
+    public Void visitMethodInvocation(MethodInvocationTree node, Void unused) {
+      checkIgnoredExpression(node);
+      return super.visitMethodInvocation(node, unused);
+    }
+
+    @Override
+    public Void visitNewClass(NewClassTree node, Void unused) {
+      checkIgnoredExpression(node);
+      return super.visitNewClass(node, unused);
+    }
+
+    /**
+     * Common logic for both method invocations and constructor calls when they appear as
+     * stand-alone expression statements (i.e., their result is ignored).
+     */
+    private void checkIgnoredExpression(Tree node) {
+      var path = m_trees.getPath(m_root, node);
+      var parentPath = (path == null) ? null : path.getParentPath();
+      if (parentPath == null || parentPath.getLeaf().getKind() != Tree.Kind.EXPRESSION_STATEMENT) {
+        // If the parent node is an expression statement, then the return value is ignored.
+        // Otherwise, the return value is used and we can ignore this site.
+        return;
+      }
+
+      // Resolve the static type of the expression
+      TypeMirror returnType = getType(node);
+      if (returnType == null || returnType.getKind() == TypeKind.VOID) {
+        // Skip void since there's nothing to return.
+        // Maybe this should be a compiler error, but it might be bad library code and we shouldn't
+        // have it break users.
+        return;
+      }
+
+      m_specialTypes.forEach(
+          (type, msg) -> {
+            if (type == null) {
+              // This type is not on the classpath in this compilation unit (i.e., users are not
+              // using it); skip.
+              return;
+            }
+
+            // Also applies to subtypes
+            boolean isSpecial = m_task.getTypes().isAssignable(returnType, type.asType());
+            if (isSpecial) {
+              m_trees.printMessage(Diagnostic.Kind.ERROR, msg, node, m_root);
+            }
+          });
+
+      // Check @NoDiscard on the invoked executable (method or constructor)
+      var invoked = getInvokedExecutable(node);
+      if (invoked != null) {
+        String msg = getNoDiscardMessage(invoked);
+        if (msg != null) {
+          m_trees.printMessage(Diagnostic.Kind.ERROR, msg, node, m_root);
+        }
+      }
+    }
+
+    private TypeMirror getType(Tree node) {
+      var path = m_trees.getPath(m_root, node);
+      if (path == null) {
+        return null;
+      }
+      // Requires running after ANALYZE attribution has completed for this CU.
+      return m_trees.getTypeMirror(path);
+    }
+
+    private ExecutableElement getInvokedExecutable(Tree node) {
+      var path = m_trees.getPath(m_root, node);
+      if (path == null) {
+        return null;
+      }
+      var el = m_trees.getElement(path);
+      return (el instanceof ExecutableElement ee) ? ee : null;
+    }
+
+    private String getNoDiscardMessage(ExecutableElement method) {
+      for (var mirror : method.getAnnotationMirrors()) {
```

Comment:
I'm curious - does TypeMirror.getAnnotation(NoDiscard.class) not work here?

Reply:
The annotation project isn't on the classpath, but there's no real reason it couldn't be added